Src/VimWpf/RelativeLineNumbers/LineNumbersCalculator.cs

```diff
@@ -0,0 +1,106 @@
+﻿using System;
+using System.Collections.Generic;
+using System.Linq;
+
+using Microsoft.VisualStudio.Text;
+using Microsoft.VisualStudio.Text.Editor;
+using Microsoft.VisualStudio.Text.Formatting;
+
+using Vim.UI.Wpf.RelativeLineNumbers.Util;
+
+namespace Vim.UI.Wpf.RelativeLineNumbers
+{
+    public class LineNumbersCalculator
+    {
+        private readonly IWpfTextView _textView;
+
+        private readonly ILineFormatTracker _formatTracker;
+
+        public LineNumbersCalculator(IWpfTextView textView, ILineFormatTracker formatTracker)
+        {
+            _textView = textView;
+            _formatTracker = formatTracker;
+        }
+
+        public ICollection<Line> CalculateLineNumbers()
+        {
+            int caretIndex = GetCaretIndex();
+
+            var result = GetLinesWithNumbers()
+                .Select((line, idx) => MakeLine(line, idx, caretIndex))
+                .ToList();
+
+            return result;
+        }
+
+        private IEnumerable<ITextViewLine> GetLinesWithNumbers()
+        {
+            var allLines = _textView.TextViewLines;
+
+            return allLines.Where(x => x.IsFirstTextViewLineForSnapshotLine);
+        }
+
+        private int GetCaretIndex()
+        {
+            var caretLine = _textView.Caret.ContainingTextViewLine;
+
+            var firstVisibleLine =
+                _textView.TextViewLines.First(x => x.IsFirstTextViewLineForSnapshotLine);

```

Comment:
The `TextViewLines` property can throw if we're in the middle of rendering already. Are we sufficiently guarded from that happening here?

Reply:
We should be safe here as this is called in the UI thread and LineNumbersTracker encapsulates all the logic for deciding when we should trigger actions on "NumbersChange". It also ensures that it only fires the event if "InLayout" is false for the TextView (as well as "IsClosed").

Reply:
As it turns out, `ContainingTextViewLine` can throw as well. The new PR has a fix for this.